language/functional_tests/src/evaluator.rs

```diff
@@ -81,44 +82,46 @@ impl OutputType {
     }
 }
 
-/// An entry in the `EvaluationResult`.
+/// An entry in the `EvaluationLog`.
 #[derive(Debug)]
 pub enum EvaluationOutput {
-    Transaction,
+    Transaction(usize),
     Stage(Stage),
     Output(Box<OutputType>),
-    Error(String),
+    Error(Box<Error>),
+    Status(Status),
 }
 
 /// A log consisting of outputs from all stages and the final status.
 /// This is checked against the directives.
 #[derive(Debug)]
-pub struct EvaluationResult {
+pub struct EvaluationLog {
     pub outputs: Vec<EvaluationOutput>,
-    pub status: Status,
-    pub use_debug_output: bool,
 }
 
-impl EvaluationResult {
-    pub fn get_transaction_count(&self) -> usize {
-        self.outputs
-            .iter()
-            .filter(|output| match output {
-                EvaluationOutput::Transaction => true,
-                _ => false,
-            })
-            .count()
-    }
-
-    pub fn get_last_stage(&self) -> Option<Stage> {
-        let mut stage = None;
-        for output in self.outputs.iter().rev() {
-            if let EvaluationOutput::Stage(s) = output {
-                stage = Some(*s);
-                break;
+impl EvaluationLog {
+    pub fn get_failed_transactions(&self) -> Vec<(usize, Stage)> {
+        let mut res = vec![];
+        let mut last_txn = None;
+        let mut last_stage = None;
+
+        for output in &self.outputs {
+            match output {
+                EvaluationOutput::Transaction(idx) => last_txn = Some(idx),
+                EvaluationOutput::Stage(stage) => last_stage = Some(stage),
+                EvaluationOutput::Status(Status::Failure) => match (last_txn, last_stage) {
+                    (Some(idx), Some(stage)) => res.push((*idx, *stage)),
+                    _ => unreachable!(),
+                },
+                _ => (),
             }
         }
-        stage
+
+        res
+    }
+
+    fn append(&mut self, output: EvaluationOutput) {
```

Comment:
It seems like we should either make `outputs` private (preferred) or eliminate this and let the caller do `outputs.push()`?

Reply:
Sounds reasonable. I'll make the field private.

Reply:
`outputs` is used in the checker. I'm planning to refactor that part pretty aggressively and it would be much easier for me to determine the API until then. I'd keep the field public for now.